File: algorithms/easy/minimum_number_of_operations_to_make_elements_in_array_distinct.py
from typing import List

class Solution:
    def minimumOperations(self, nums: List[int]) -> int:
        # Brute force; check the remainder of the array after each jump of 3 steps.
        N = len(nums)
        for i in range(0, N + 3, 3):
            if len(set(nums[i:])) == len(nums[i:]):
                return i // 3

# Main section
for nums in [
               [1,2,3,4,2,3,3,5,7],
               [4,5,6,4,4],
               [6,7,8,9],
            ]:
    print(f'nums = {nums}')
    sol = Solution()
    r = sol.minimumOperations(nums)
    print(f'r = {r}')
    print('========================')

File: algorithms/easy/minimum_number_of_operations_to_make_elements_in_array_distinct.go
package main

import "fmt"

func minimumOperations(nums []int) int {
	seen := make(map[int]bool)
	i := len(nums) - 1
	for i >= 0 {
		ok, _ := seen[nums[i]]
		if ok {
			break
		}
		seen[nums[i]] = true
		i--
	}
	return (i + 3) / 3
}

func main() {
	data := [][]int{
		[]int{1, 2, 3, 4, 2, 3, 3, 5, 7},
		[]int{4, 5, 6, 4, 4},
		[]int{6, 7, 8, 9},
	}
	for _, nums := range data {
		fmt.Printf("nums = %v\n", nums)
		r := minimumOperations(nums)
		fmt.Printf("r = %d\n", r)
		fmt.Println("====================")
	}
}
